diff --git a/Sources/ArgumentParserTestHelpers/TestHelpers.swift b/Sources/ArgumentParserTestHelpers/TestHelpers.swift
index 33d5638da..debcfd67c 100644
--- a/Sources/ArgumentParserTestHelpers/TestHelpers.swift
+++ b/Sources/ArgumentParserTestHelpers/TestHelpers.swift
@@ -389,14 +389,17 @@ extension XCTest {
     test: StaticString = #function,
     file: StaticString = #filePath,
     line: UInt = #line
-  ) throws -> String {
+  ) throws -> String? {
     let snapshotDirectoryURL = URL(fileURLWithPath: "\(file)")
       .deletingLastPathComponent()
       .appendingPathComponent("Snapshots")
     let snapshotFileURL = snapshotDirectoryURL
       .appendingPathComponent("\(test).\(`extension`)")
 
-    if record || !FileManager.default.fileExists(atPath: snapshotFileURL.path) {
+    let snapshotExists = FileManager.default.fileExists(atPath: snapshotFileURL.path)
+    let recordEnvironment = ProcessInfo.processInfo.environment["RECORD_SNAPSHOTS"] != nil
+
+    if record || recordEnvironment || !snapshotExists {
       let recordedValue = actual + "\n"
       try FileManager.default.createDirectory(
         at: snapshotDirectoryURL,
@@ -404,8 +407,7 @@ extension XCTest {
         attributes: nil)
       try recordedValue.write(to: snapshotFileURL, atomically: true, encoding: .utf8)
       XCTFail("Recorded new baseline", file: file, line: line)
-      struct EarlyExit: Error {}
-      throw EarlyExit()
+      return nil
     } else {
       let expected = try String(contentsOf: snapshotFileURL, encoding: .utf8)
       AssertEqualStrings(
@@ -513,6 +515,8 @@ extension XCTest {
       file: file,
       line: line)
 
+    guard let expected else { return }
+
     try AssertJSONEqualFromString(
       actual: actual,
       expected: expected,
diff --git a/Tests/ArgumentParserExampleTests/MathExampleTests.swift b/Tests/ArgumentParserExampleTests/MathExampleTests.swift
index ba191f928..360977e26 100644
--- a/Tests/ArgumentParserExampleTests/MathExampleTests.swift
+++ b/Tests/ArgumentParserExampleTests/MathExampleTests.swift
@@ -190,27 +190,24 @@ final class MathExampleTests: XCTestCase {
 // MARK: - Completion Script
 
 extension MathExampleTests {
-  func testMath_CompletionScript() throws {
-    try AssertExecuteCommand(
-      command: "math --generate-completion-script=bash",
-      expected: bashCompletionScriptText)
-    try AssertExecuteCommand(
-      command: "math --generate-completion-script bash",
-      expected: bashCompletionScriptText)
-    try AssertExecuteCommand(
-      command: "math --generate-completion-script=zsh",
-      expected: zshCompletionScriptText)
-    try AssertExecuteCommand(
-      command: "math --generate-completion-script zsh",
-      expected: zshCompletionScriptText)
-    try AssertExecuteCommand(
-      command: "math --generate-completion-script=fish",
-      expected: fishCompletionScriptText)
-    try AssertExecuteCommand(
-      command: "math --generate-completion-script fish",
-      expected: fishCompletionScriptText)
+  func testMathBashCompletionScript() throws {
+    let script = try AssertExecuteCommand(
+      command: "math --generate-completion-script bash")
+    try assertSnapshot(actual: script, extension: "bash")
   }
-  
+
+  func testMathZshCompletionScript() throws {
+    let script = try AssertExecuteCommand(
+      command: "math --generate-completion-script zsh")
+    try assertSnapshot(actual: script, extension: "zsh")
+  }
+
+  func testMathFishCompletionScript() throws {
+    let script = try AssertExecuteCommand(
+      command: "math --generate-completion-script fish")
+    try assertSnapshot(actual: script, extension: "fish")
+  }
+
   func testMath_CustomCompletion() throws {
     try AssertExecuteCommand(
       command: "math ---completion stats quantiles -- --custom",
@@ -236,414 +233,3 @@ extension MathExampleTests {
         """)
   }
 }
-
-private let bashCompletionScriptText = """
-#!/bin/bash
-
-_math() {
-    export SAP_SHELL=bash
-    SAP_SHELL_VERSION="$(IFS='.'; printf %s "${BASH_VERSINFO[*]}")"
-    export SAP_SHELL_VERSION
-    cur="${COMP_WORDS[COMP_CWORD]}"
-    prev="${COMP_WORDS[COMP_CWORD-1]}"
-    COMPREPLY=()
-    opts="--version -h --help add multiply stats help"
-    if [[ $COMP_CWORD == "1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    case ${COMP_WORDS[1]} in
-        (add)
-            _math_add 2
-            return
-            ;;
-        (multiply)
-            _math_multiply 2
-            return
-            ;;
-        (stats)
-            _math_stats 2
-            return
-            ;;
-        (help)
-            _math_help 2
-            return
-            ;;
-    esac
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-_math_add() {
-    opts="--hex-output -x --version -h --help"
-    if [[ $COMP_CWORD == "$1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-_math_multiply() {
-    opts="--hex-output -x --version -h --help"
-    if [[ $COMP_CWORD == "$1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-_math_stats() {
-    opts="--version -h --help average stdev quantiles"
-    if [[ $COMP_CWORD == "$1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    case ${COMP_WORDS[$1]} in
-        (average)
-            _math_stats_average $(($1+1))
-            return
-            ;;
-        (stdev)
-            _math_stats_stdev $(($1+1))
-            return
-            ;;
-        (quantiles)
-            _math_stats_quantiles $(($1+1))
-            return
-            ;;
-    esac
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-_math_stats_average() {
-    opts="--kind --version -h --help"
-    if [[ $COMP_CWORD == "$1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    case $prev in
-        --kind)
-            COMPREPLY=( $(compgen -W "mean median mode" -- "$cur") )
-            return
-        ;;
-    esac
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-_math_stats_stdev() {
-    opts="--version -h --help"
-    if [[ $COMP_CWORD == "$1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-_math_stats_quantiles() {
-    opts="--file --directory --shell --custom --version -h --help"
-    opts="$opts alphabet alligator branch braggart"
-    opts="$opts $("${COMP_WORDS[0]}" ---completion stats quantiles -- customArg "${COMP_WORDS[@]}")"
-    if [[ $COMP_CWORD == "$1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    case $prev in
-        --file)
-            if declare -F _filedir >/dev/null; then
-              _filedir 'txt'
-              _filedir 'md'
-              _filedir 'TXT'
-              _filedir 'MD'
-              _filedir -d
-            else
-              COMPREPLY=(
-                $(compgen -f -X '!*.txt' -- "$cur")
-                $(compgen -f -X '!*.md' -- "$cur")
-                $(compgen -f -X '!*.TXT' -- "$cur")
-                $(compgen -f -X '!*.MD' -- "$cur")
-                $(compgen -d -- "$cur")
-              )
-            fi
-            return
-        ;;
-        --directory)
-            if declare -F _filedir >/dev/null; then
-              _filedir -d
-            else
-              COMPREPLY=( $(compgen -d -- "$cur") )
-            fi
-            return
-        ;;
-        --shell)
-            COMPREPLY=( $(head -100 /usr/share/dict/words | tail -50) )
-            return
-        ;;
-        --custom)
-            COMPREPLY=( $(compgen -W "$("${COMP_WORDS[0]}" ---completion stats quantiles -- --custom "${COMP_WORDS[@]}")" -- "$cur") )
-            return
-        ;;
-    esac
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-_math_help() {
-    opts="--version"
-    if [[ $COMP_CWORD == "$1" ]]; then
-        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-        return
-    fi
-    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
-}
-
-
-complete -F _math math
-"""
-
-private let zshCompletionScriptText = """
-#compdef math
-local context state state_descr line
-_math_commandname=$words[1]
-typeset -A opt_args
-
-_math() {
-    export SAP_SHELL=zsh
-    SAP_SHELL_VERSION="$(builtin emulate zsh -c 'printf %s "${ZSH_VERSION}"')"
-    export SAP_SHELL_VERSION
-    integer ret=1
-    local -a args
-    args+=(
-        '--version[Show the version.]'
-        '(-h --help)'{-h,--help}'[Show help information.]'
-        '(-): :->command'
-        '(-)*:: :->arg'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-    case $state in
-        (command)
-            local subcommands
-            subcommands=(
-                'add:Print the sum of the values.'
-                'multiply:Print the product of the values.'
-                'stats:Calculate descriptive statistics.'
-                'help:Show subcommand help information.'
-            )
-            _describe "subcommand" subcommands
-            ;;
-        (arg)
-            case ${words[1]} in
-                (add)
-                    _math_add
-                    ;;
-                (multiply)
-                    _math_multiply
-                    ;;
-                (stats)
-                    _math_stats
-                    ;;
-                (help)
-                    _math_help
-                    ;;
-            esac
-            ;;
-    esac
-
-    return ret
-}
-
-_math_add() {
-    integer ret=1
-    local -a args
-    args+=(
-        '(--hex-output -x)'{--hex-output,-x}'[Use hexadecimal notation for the result.]'
-        ':values:'
-        '--version[Show the version.]'
-        '(-h --help)'{-h,--help}'[Show help information.]'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-
-    return ret
-}
-
-_math_multiply() {
-    integer ret=1
-    local -a args
-    args+=(
-        '(--hex-output -x)'{--hex-output,-x}'[Use hexadecimal notation for the result.]'
-        ':values:'
-        '--version[Show the version.]'
-        '(-h --help)'{-h,--help}'[Show help information.]'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-
-    return ret
-}
-
-_math_stats() {
-    integer ret=1
-    local -a args
-    args+=(
-        '--version[Show the version.]'
-        '(-h --help)'{-h,--help}'[Show help information.]'
-        '(-): :->command'
-        '(-)*:: :->arg'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-    case $state in
-        (command)
-            local subcommands
-            subcommands=(
-                'average:Print the average of the values.'
-                'stdev:Print the standard deviation of the values.'
-                'quantiles:Print the quantiles of the values (TBD).'
-            )
-            _describe "subcommand" subcommands
-            ;;
-        (arg)
-            case ${words[1]} in
-                (average)
-                    _math_stats_average
-                    ;;
-                (stdev)
-                    _math_stats_stdev
-                    ;;
-                (quantiles)
-                    _math_stats_quantiles
-                    ;;
-            esac
-            ;;
-    esac
-
-    return ret
-}
-
-_math_stats_average() {
-    integer ret=1
-    local -a args
-    args+=(
-        '--kind[The kind of average to provide.]:kind:(mean median mode)'
-        ':values:'
-        '--version[Show the version.]'
-        '(-h --help)'{-h,--help}'[Show help information.]'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-
-    return ret
-}
-
-_math_stats_stdev() {
-    integer ret=1
-    local -a args
-    args+=(
-        ':values:'
-        '--version[Show the version.]'
-        '(-h --help)'{-h,--help}'[Show help information.]'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-
-    return ret
-}
-
-_math_stats_quantiles() {
-    integer ret=1
-    local -a args
-    args+=(
-        ':one-of-four:(alphabet alligator branch braggart)'
-        ':custom-arg:{_custom_completion $_math_commandname ---completion stats quantiles -- customArg $words}'
-        ':values:'
-        '--file:file:_files -g '"'"'*.txt *.md'"'"''
-        '--directory:directory:_files -/'
-        '--shell:shell:{local -a list; list=(${(f)"$(head -100 /usr/share/dict/words | tail -50)"}); _describe '''' list}'
-        '--custom:custom:{_custom_completion $_math_commandname ---completion stats quantiles -- --custom $words}'
-        '--version[Show the version.]'
-        '(-h --help)'{-h,--help}'[Show help information.]'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-
-    return ret
-}
-
-_math_help() {
-    integer ret=1
-    local -a args
-    args+=(
-        ':subcommands:'
-        '--version[Show the version.]'
-    )
-    _arguments -w -s -S $args[@] && ret=0
-
-    return ret
-}
-
-
-_custom_completion() {
-    local completions=("${(@f)$($*)}")
-    _describe '' completions
-}
-
-_math
-"""
-
-private let fishCompletionScriptText = """
-# A function which filters options which starts with \"-\" from $argv.
-function _swift_math_preprocessor
-    set -l results
-    for i in (seq (count $argv))
-        switch (echo $argv[$i] | string sub -l 1)
-            case '-'
-            case '*'
-                echo $argv[$i]
-        end
-    end
-end
-
-function _swift_math_using_command
-    set -gx SAP_SHELL fish
-    set -gx SAP_SHELL_VERSION "$FISH_VERSION"
-    set -l currentCommands (_swift_math_preprocessor (commandline -opc))
-    set -l expectedCommands (string split \" \" $argv[1])
-    set -l subcommands (string split \" \" $argv[2])
-    if [ (count $currentCommands) -ge (count $expectedCommands) ]
-        for i in (seq (count $expectedCommands))
-            if [ $currentCommands[$i] != $expectedCommands[$i] ]
-                return 1
-            end
-        end
-        if [ (count $currentCommands) -eq (count $expectedCommands) ]
-            return 0
-        end
-        if [ (count $subcommands) -gt 1 ]
-            for i in (seq (count $subcommands))
-                if [ $currentCommands[(math (count $expectedCommands) + 1)] = $subcommands[$i] ]
-                    return 1
-                end
-            end
-        end
-        return 0
-    end
-    return 1
-end
-
-complete -c math -n \'_swift_math_using_command \"math add\"\' -l hex-output -s x -d \'Use hexadecimal notation for the result.\'
-complete -c math -n \'_swift_math_using_command \"math add\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math add\"\' -s h -l help -d \'Show help information.\'
-complete -c math -n \'_swift_math_using_command \"math multiply\"\' -l hex-output -s x -d \'Use hexadecimal notation for the result.\'
-complete -c math -n \'_swift_math_using_command \"math multiply\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math multiply\"\' -s h -l help -d \'Show help information.\'
-complete -c math -n \'_swift_math_using_command \"math stats average\"\' -l kind -d \'The kind of average to provide.\' -r -f -k -a \'mean median mode\'
-complete -c math -n \'_swift_math_using_command \"math stats average\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math stats average\"\' -s h -l help -d \'Show help information.\'
-complete -c math -n \'_swift_math_using_command \"math stats stdev\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math stats stdev\"\' -s h -l help -d \'Show help information.\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -r -f -k -a \'alphabet alligator branch braggart\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -r -f -a \'(command math ---completion stats quantiles -- customArg (commandline -opc)[1..-1])\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -l file -r -f -a \'(for i in *.{txt,md}; echo $i;end)\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -l directory -r -f -a \'(__fish_complete_directories)\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -l shell -r -f -a \'(head -100 /usr/share/dict/words | tail -50)\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -l custom -r -f -a \'(command math ---completion stats quantiles -- --custom (commandline -opc)[1..-1])\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math stats quantiles\"\' -s h -l help -d \'Show help information.\'
-complete -c math -n \'_swift_math_using_command \"math stats\" \"average stdev quantiles\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math stats\" \"average stdev quantiles\"\' -s h -l help -d \'Show help information.\'
-complete -c math -n \'_swift_math_using_command \"math stats\" \"average stdev quantiles\"\' -f -a \'average\' -d \'Print the average of the values.\'
-complete -c math -n \'_swift_math_using_command \"math stats\" \"average stdev quantiles\"\' -f -a \'stdev\' -d \'Print the standard deviation of the values.\'
-complete -c math -n \'_swift_math_using_command \"math stats\" \"average stdev quantiles\"\' -f -a \'quantiles\' -d \'Print the quantiles of the values (TBD).\'
-complete -c math -n \'_swift_math_using_command \"math help\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math\" \"add multiply stats help\"\' -l version -d \'Show the version.\'
-complete -c math -n \'_swift_math_using_command \"math\" \"add multiply stats help\"\' -s h -l help -d \'Show help information.\'
-complete -c math -n \'_swift_math_using_command \"math\" \"add multiply stats help\"\' -f -a \'add\' -d \'Print the sum of the values.\'
-complete -c math -n \'_swift_math_using_command \"math\" \"add multiply stats help\"\' -f -a \'multiply\' -d \'Print the product of the values.\'
-complete -c math -n \'_swift_math_using_command \"math\" \"add multiply stats help\"\' -f -a \'stats\' -d \'Calculate descriptive statistics.\'
-complete -c math -n \'_swift_math_using_command \"math\" \"add multiply stats help\"\' -f -a \'help\' -d \'Show subcommand help information.\'
-"""
diff --git a/Tests/ArgumentParserExampleTests/Snapshots/testMathBashCompletionScript().bash b/Tests/ArgumentParserExampleTests/Snapshots/testMathBashCompletionScript().bash
new file mode 100644
index 000000000..815cbed80
--- /dev/null
+++ b/Tests/ArgumentParserExampleTests/Snapshots/testMathBashCompletionScript().bash
@@ -0,0 +1,151 @@
+#!/bin/bash
+
+_math() {
+    export SAP_SHELL=bash
+    SAP_SHELL_VERSION="$(IFS='.'; printf %s "${BASH_VERSINFO[*]}")"
+    export SAP_SHELL_VERSION
+    cur="${COMP_WORDS[COMP_CWORD]}"
+    prev="${COMP_WORDS[COMP_CWORD-1]}"
+    COMPREPLY=()
+    opts="--version -h --help add multiply stats help"
+    if [[ $COMP_CWORD == "1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    case ${COMP_WORDS[1]} in
+        (add)
+            _math_add 2
+            return
+            ;;
+        (multiply)
+            _math_multiply 2
+            return
+            ;;
+        (stats)
+            _math_stats 2
+            return
+            ;;
+        (help)
+            _math_help 2
+            return
+            ;;
+    esac
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+_math_add() {
+    opts="--hex-output -x --version -h --help"
+    if [[ $COMP_CWORD == "$1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+_math_multiply() {
+    opts="--hex-output -x --version -h --help"
+    if [[ $COMP_CWORD == "$1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+_math_stats() {
+    opts="--version -h --help average stdev quantiles"
+    if [[ $COMP_CWORD == "$1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    case ${COMP_WORDS[$1]} in
+        (average)
+            _math_stats_average $(($1+1))
+            return
+            ;;
+        (stdev)
+            _math_stats_stdev $(($1+1))
+            return
+            ;;
+        (quantiles)
+            _math_stats_quantiles $(($1+1))
+            return
+            ;;
+    esac
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+_math_stats_average() {
+    opts="--kind --version -h --help"
+    if [[ $COMP_CWORD == "$1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    case $prev in
+        --kind)
+            COMPREPLY=( $(compgen -W "mean median mode" -- "$cur") )
+            return
+        ;;
+    esac
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+_math_stats_stdev() {
+    opts="--version -h --help"
+    if [[ $COMP_CWORD == "$1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+_math_stats_quantiles() {
+    opts="--file --directory --shell --custom --version -h --help"
+    opts="$opts alphabet alligator branch braggart"
+    opts="$opts $("${COMP_WORDS[0]}" ---completion stats quantiles -- customArg "${COMP_WORDS[@]}")"
+    if [[ $COMP_CWORD == "$1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    case $prev in
+        --file)
+            if declare -F _filedir >/dev/null; then
+              _filedir 'txt'
+              _filedir 'md'
+              _filedir 'TXT'
+              _filedir 'MD'
+              _filedir -d
+            else
+              COMPREPLY=(
+                $(compgen -f -X '!*.txt' -- "$cur")
+                $(compgen -f -X '!*.md' -- "$cur")
+                $(compgen -f -X '!*.TXT' -- "$cur")
+                $(compgen -f -X '!*.MD' -- "$cur")
+                $(compgen -d -- "$cur")
+              )
+            fi
+            return
+        ;;
+        --directory)
+            if declare -F _filedir >/dev/null; then
+              _filedir -d
+            else
+              COMPREPLY=( $(compgen -d -- "$cur") )
+            fi
+            return
+        ;;
+        --shell)
+            COMPREPLY=( $(head -100 /usr/share/dict/words | tail -50) )
+            return
+        ;;
+        --custom)
+            COMPREPLY=( $(compgen -W "$("${COMP_WORDS[0]}" ---completion stats quantiles -- --custom "${COMP_WORDS[@]}")" -- "$cur") )
+            return
+        ;;
+    esac
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+_math_help() {
+    opts="--version"
+    if [[ $COMP_CWORD == "$1" ]]; then
+        COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+        return
+    fi
+    COMPREPLY=( $(compgen -W "$opts" -- "$cur") )
+}
+
+
+complete -F _math math
diff --git a/Tests/ArgumentParserExampleTests/Snapshots/testMathFishCompletionScript().fish b/Tests/ArgumentParserExampleTests/Snapshots/testMathFishCompletionScript().fish
new file mode 100644
index 000000000..d109f9410
--- /dev/null
+++ b/Tests/ArgumentParserExampleTests/Snapshots/testMathFishCompletionScript().fish
@@ -0,0 +1,70 @@
+# A function which filters options which starts with "-" from $argv.
+function _swift_math_preprocessor
+    set -l results
+    for i in (seq (count $argv))
+        switch (echo $argv[$i] | string sub -l 1)
+            case '-'
+            case '*'
+                echo $argv[$i]
+        end
+    end
+end
+
+function _swift_math_using_command
+    set -gx SAP_SHELL fish
+    set -gx SAP_SHELL_VERSION "$FISH_VERSION"
+    set -l currentCommands (_swift_math_preprocessor (commandline -opc))
+    set -l expectedCommands (string split " " $argv[1])
+    set -l subcommands (string split " " $argv[2])
+    if [ (count $currentCommands) -ge (count $expectedCommands) ]
+        for i in (seq (count $expectedCommands))
+            if [ $currentCommands[$i] != $expectedCommands[$i] ]
+                return 1
+            end
+        end
+        if [ (count $currentCommands) -eq (count $expectedCommands) ]
+            return 0
+        end
+        if [ (count $subcommands) -gt 1 ]
+            for i in (seq (count $subcommands))
+                if [ $currentCommands[(math (count $expectedCommands) + 1)] = $subcommands[$i] ]
+                    return 1
+                end
+            end
+        end
+        return 0
+    end
+    return 1
+end
+
+complete -c math -n '_swift_math_using_command "math add"' -l hex-output -s x -d 'Use hexadecimal notation for the result.'
+complete -c math -n '_swift_math_using_command "math add"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math add"' -s h -l help -d 'Show help information.'
+complete -c math -n '_swift_math_using_command "math multiply"' -l hex-output -s x -d 'Use hexadecimal notation for the result.'
+complete -c math -n '_swift_math_using_command "math multiply"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math multiply"' -s h -l help -d 'Show help information.'
+complete -c math -n '_swift_math_using_command "math stats average"' -l kind -d 'The kind of average to provide.' -r -f -k -a 'mean median mode'
+complete -c math -n '_swift_math_using_command "math stats average"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math stats average"' -s h -l help -d 'Show help information.'
+complete -c math -n '_swift_math_using_command "math stats stdev"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math stats stdev"' -s h -l help -d 'Show help information.'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -r -f -k -a 'alphabet alligator branch braggart'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -r -f -a '(command math ---completion stats quantiles -- customArg (commandline -opc)[1..-1])'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -l file -r -f -a '(for i in *.{txt,md}; echo $i;end)'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -l directory -r -f -a '(__fish_complete_directories)'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -l shell -r -f -a '(head -100 /usr/share/dict/words | tail -50)'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -l custom -r -f -a '(command math ---completion stats quantiles -- --custom (commandline -opc)[1..-1])'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math stats quantiles"' -s h -l help -d 'Show help information.'
+complete -c math -n '_swift_math_using_command "math stats" "average stdev quantiles"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math stats" "average stdev quantiles"' -s h -l help -d 'Show help information.'
+complete -c math -n '_swift_math_using_command "math stats" "average stdev quantiles"' -f -a 'average' -d 'Print the average of the values.'
+complete -c math -n '_swift_math_using_command "math stats" "average stdev quantiles"' -f -a 'stdev' -d 'Print the standard deviation of the values.'
+complete -c math -n '_swift_math_using_command "math stats" "average stdev quantiles"' -f -a 'quantiles' -d 'Print the quantiles of the values (TBD).'
+complete -c math -n '_swift_math_using_command "math help"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math" "add multiply stats help"' -l version -d 'Show the version.'
+complete -c math -n '_swift_math_using_command "math" "add multiply stats help"' -s h -l help -d 'Show help information.'
+complete -c math -n '_swift_math_using_command "math" "add multiply stats help"' -f -a 'add' -d 'Print the sum of the values.'
+complete -c math -n '_swift_math_using_command "math" "add multiply stats help"' -f -a 'multiply' -d 'Print the product of the values.'
+complete -c math -n '_swift_math_using_command "math" "add multiply stats help"' -f -a 'stats' -d 'Calculate descriptive statistics.'
+complete -c math -n '_swift_math_using_command "math" "add multiply stats help"' -f -a 'help' -d 'Show subcommand help information.'
diff --git a/Tests/ArgumentParserExampleTests/Snapshots/testMathZshCompletionScript().zsh b/Tests/ArgumentParserExampleTests/Snapshots/testMathZshCompletionScript().zsh
new file mode 100644
index 000000000..ba5b1db74
--- /dev/null
+++ b/Tests/ArgumentParserExampleTests/Snapshots/testMathZshCompletionScript().zsh
@@ -0,0 +1,181 @@
+#compdef math
+local context state state_descr line
+_math_commandname=$words[1]
+typeset -A opt_args
+
+_math() {
+    export SAP_SHELL=zsh
+    SAP_SHELL_VERSION="$(builtin emulate zsh -c 'printf %s "${ZSH_VERSION}"')"
+    export SAP_SHELL_VERSION
+    integer ret=1
+    local -a args
+    args+=(
+        '--version[Show the version.]'
+        '(-h --help)'{-h,--help}'[Show help information.]'
+        '(-): :->command'
+        '(-)*:: :->arg'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+    case $state in
+        (command)
+            local subcommands
+            subcommands=(
+                'add:Print the sum of the values.'
+                'multiply:Print the product of the values.'
+                'stats:Calculate descriptive statistics.'
+                'help:Show subcommand help information.'
+            )
+            _describe "subcommand" subcommands
+            ;;
+        (arg)
+            case ${words[1]} in
+                (add)
+                    _math_add
+                    ;;
+                (multiply)
+                    _math_multiply
+                    ;;
+                (stats)
+                    _math_stats
+                    ;;
+                (help)
+                    _math_help
+                    ;;
+            esac
+            ;;
+    esac
+
+    return ret
+}
+
+_math_add() {
+    integer ret=1
+    local -a args
+    args+=(
+        '(--hex-output -x)'{--hex-output,-x}'[Use hexadecimal notation for the result.]'
+        ':values:'
+        '--version[Show the version.]'
+        '(-h --help)'{-h,--help}'[Show help information.]'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+
+    return ret
+}
+
+_math_multiply() {
+    integer ret=1
+    local -a args
+    args+=(
+        '(--hex-output -x)'{--hex-output,-x}'[Use hexadecimal notation for the result.]'
+        ':values:'
+        '--version[Show the version.]'
+        '(-h --help)'{-h,--help}'[Show help information.]'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+
+    return ret
+}
+
+_math_stats() {
+    integer ret=1
+    local -a args
+    args+=(
+        '--version[Show the version.]'
+        '(-h --help)'{-h,--help}'[Show help information.]'
+        '(-): :->command'
+        '(-)*:: :->arg'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+    case $state in
+        (command)
+            local subcommands
+            subcommands=(
+                'average:Print the average of the values.'
+                'stdev:Print the standard deviation of the values.'
+                'quantiles:Print the quantiles of the values (TBD).'
+            )
+            _describe "subcommand" subcommands
+            ;;
+        (arg)
+            case ${words[1]} in
+                (average)
+                    _math_stats_average
+                    ;;
+                (stdev)
+                    _math_stats_stdev
+                    ;;
+                (quantiles)
+                    _math_stats_quantiles
+                    ;;
+            esac
+            ;;
+    esac
+
+    return ret
+}
+
+_math_stats_average() {
+    integer ret=1
+    local -a args
+    args+=(
+        '--kind[The kind of average to provide.]:kind:(mean median mode)'
+        ':values:'
+        '--version[Show the version.]'
+        '(-h --help)'{-h,--help}'[Show help information.]'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+
+    return ret
+}
+
+_math_stats_stdev() {
+    integer ret=1
+    local -a args
+    args+=(
+        ':values:'
+        '--version[Show the version.]'
+        '(-h --help)'{-h,--help}'[Show help information.]'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+
+    return ret
+}
+
+_math_stats_quantiles() {
+    integer ret=1
+    local -a args
+    args+=(
+        ':one-of-four:(alphabet alligator branch braggart)'
+        ':custom-arg:{_custom_completion $_math_commandname ---completion stats quantiles -- customArg $words}'
+        ':values:'
+        '--file:file:_files -g '"'"'*.txt *.md'"'"''
+        '--directory:directory:_files -/'
+        '--shell:shell:{local -a list; list=(${(f)"$(head -100 /usr/share/dict/words | tail -50)"}); _describe '''' list}'
+        '--custom:custom:{_custom_completion $_math_commandname ---completion stats quantiles -- --custom $words}'
+        '--version[Show the version.]'
+        '(-h --help)'{-h,--help}'[Show help information.]'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+
+    return ret
+}
+
+_math_help() {
+    integer ret=1
+    local -a args
+    args+=(
+        ':subcommands:'
+        '--version[Show the version.]'
+    )
+    _arguments -w -s -S $args[@] && ret=0
+
+    return ret
+}
+
+
+_custom_completion() {
+    local completions=("${(@f)$($*)}")
+    _describe '' completions
+}
+
+_math
